File: python/workers/gpu_worker.py
import asyncio
import logging
import os
import sys

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from temporalio.client import Client
from temporalio.worker import Worker

from activities.gpu import GPUBoundActivities


async def main():
    """Run the GPU-bound activity worker"""
    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)

    # Create Temporal client
    temporal_host = os.environ.get("TEMPORAL_HOST", "localhost:7233")
    client = await Client.connect(temporal_host)

    # Get GPU count from environment
    gpu_count = int(os.environ.get("GPU_COUNT", "2"))

    # Create GPU activities instance
    gpu_activities = GPUBoundActivities(gpu_count=gpu_count)

    # Create worker for GPU-bound queue
    worker = Worker(
        client,
        task_queue="gpu-bound-queue",
        activities=[
            gpu_activities.generate_embeddings,
            gpu_activities.classify_document,
            gpu_activities.perform_ocr,
            gpu_activities.analyze_images,
            gpu_activities.run_llm_inference,
        ],
        max_concurrent_activities=gpu_count * 4,  # 4 concurrent tasks per GPU
    )

    logger.info(f"Starting GPU Worker on queue: gpu-bound-queue with {gpu_count} GPUs")
    logger.info("Optimized for GPU-accelerated ML inference and image processing")

    # Run the worker
    await worker.run()


if __name__ == "__main__":
    asyncio.run(main())

File: java/src/main/java/com/example/workers/GPUWorker.java
package com.example.workers;

import com.example.activities.gpu.GPUBoundActivitiesImpl;
import io.temporal.client.WorkflowClient;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.Worker;
import io.temporal.worker.WorkerFactory;
import io.temporal.worker.WorkerOptions;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Profile;

@Slf4j
@SpringBootApplication(scanBasePackages = "com.example")
@Profile("gpu-worker")
public class GPUWorker {
    
    private static final String TASK_QUEUE = "gpu-bound-queue";
    
    public static void main(String[] args) {
        SpringApplication.run(GPUWorker.class, args);
    }
    
    @Bean
    public CommandLineRunner run(WorkflowServiceStubs service, GPUBoundActivitiesImpl activities) {
        return args -> {
            WorkflowClient client = WorkflowClient.newInstance(service);
            WorkerFactory factory = WorkerFactory.newInstance(client);
            
            // Configure worker for GPU-intensive tasks
            // Limit concurrent activities to available GPUs
            int availableGPUs = getAvailableGPUCount();
            
            WorkerOptions workerOptions = WorkerOptions.newBuilder()
                .setMaxConcurrentActivityExecutionSize(availableGPUs) // One activity per GPU
                .setMaxConcurrentWorkflowTaskExecutionSize(50)
                .setMaxConcurrentLocalActivityExecutionSize(0) // No local activities
                .setActivityPollThreadCount(2)
                .setWorkflowPollThreadCount(1)
                .build();
            
            Worker worker = factory.newWorker(TASK_QUEUE, workerOptions);
            
            // Register GPU-bound activities
            worker.registerActivitiesImplementations(activities);
            
            // No workflows registered - this is an activity-only worker
            
            factory.start();
            
            log.info("GPU Worker started on task queue: {} with {} GPUs", 
                    TASK_QUEUE, availableGPUs);
            log.info("Optimized for GPU-accelerated ML inference and image processing");
        };
    }
    
    private int getAvailableGPUCount() {
        // In production, this would detect actual GPU devices
        // For now, simulate 2 GPUs available
        String gpuCount = System.getenv("GPU_COUNT");
        if (gpuCount != null) {
            try {
                return Integer.parseInt(gpuCount);
            } catch (NumberFormatException e) {
                log.warn("Invalid GPU_COUNT environment variable: {}", gpuCount);
            }
        }
        return 2; // Default to 2 GPUs
    }
}